### PR TITLE
State to redux

### PR DESCRIPTION
**Pull request by muselesscreator**
**TL;DR -**
Updates Gradebook to use redux to store app-related data, and clean up props.

JIRA: [JIRA-5793](https://openedx.atlassian.net/browse/JIRA-5793)
Explanation [Doc](https://docs.google.com/document/d/1jVJ5sU4a8YdiUeAwelUoRubXZWYXr6-ZVtH9zzUlV3o/edit#)

**What changed?**
Added a new *app* redux store to hold the temporary fields managed by the edit modal, as the courseId.
In removing the courseId from all of the thunkAction args (as it is now readily available to the thunkActions themselves within the redux store), I noticed that many other args were being passed through to these thunkActions that did not need to be.  Notably: the current filter state.

In particular, the fetchGrades thunkAction was highly overloaded with manual arguments for the sake of setting filter values by network response, rather than setting them locally.  Instead of setting a filter and using that to perform a fetch against current selection, we were creating a fetch manually each time, overriding only the new value, and then updating the local filter based on the api response.  The new workflow is: update the filter, and then call fetchGrades with NO arguments, as it already has access to the redux store with the current filter loadout.
Ex: `fetchGrades(courseId, selectedCohort, selectedTrack, selectedAssignmentType)` became `fetchGrades()`

A major effect of the above change is that the arguments being passed around to many components simply go away.  Many components were directly tied to not only the courseId passed ancestrally from the root Gradebook component, but also to the active selection filters, despite not accessing or doing anything special with those values.

Another minor change that came along with this was splitting out of a few components into smaller, more accessible forms for the sake of easier follow-on work: GradebookTable and EditModal each get their own directory with subcomponents, and UserLabel, SpinnerIcon, and ScoreViewInput get their own components.

* unfinished works *
I did not add unit tests for all components, but did try to maintain coverage where I could, and I believe I maintained full coverage over the redux stuff.

I also spent a bit of time going through and adding docstrings to the actions that take arguments, but did not finish, and will attempt to finish that work in the process of the upcoming unit-test-wrapup ticket.

**Developer Checklist**
- [x] Test suites passing
- [x] Documentation and test plan updated, if applicable
- [x] Received code-owner approving review
- [x] Bumped version number [package.json](../package.json)

**Testing Instructions**

[ How should a reviewer test this PR? ]

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:

- [ ] I've done a visual code review
- [x] I've tested the new functionality

FYI: @edx/masters-devs-gta

### CODE REVIEW

File: src/components/Gradebook/GradebookFilters/AssignmentGradeFilter/__snapshots__/test.jsx.snap [hunk 1]
```diff
@@ -9,14 +9,14 @@ exports[`AssignmentGradeFilter Component snapshots buttons and groups disabled i
...
-    value="1"
+    value="2"
   />
   <PercentGroup
     disabled={true}
     id="assignmentGradeMax"
     label="Max Grade"
     onChange={[MockFunction handleSetMax]}
-    value="100"
+    value="98"
```
Comment (muselesscreator) [inline src/components/Gradebook/GradebookFilters/AssignmentGradeFilter/__snapshots__/test.jsx.snap:19]: These snapshot changes are not noteworthy.  Happening because we are more-accurately mapping to the test data.

File: src/data/store.js [hunk 2]
```diff
@@ -94,5 +95,10 @@ const store = createStore(
   composeWithDevTools(applyMiddleware(thunkMiddleware, loggerMiddleware, segmentMiddleware)),
 );
 
+// TODO debug only
```
Comment (muselesscreator) [inline src/data/store.js:98]: To Remove before merge

**Review by jansenk — COMMENTED** [review]
Please try to break up PRs into smaller chunks 😢 This is very larg

**Review by nsprenkle — APPROVED** [review]
Smoke test looks good, but haven't finished visual review, hoping someone else will be able to deeper dive on visual check.

File: src/data/reducers/roles.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import actions from '../actions/roles';
 
 const initialState = {
-  canUserViewGradebook: null,
+  canUserViewGradebook: true,
```
Comment (jansenk) [inline src/data/reducers/roles.js:4]: why?

**Review by jansenk — APPROVED** [review]
I read through almost all of this and did my best to understand everything, but being honest, it's so big that something could have easily slipped through without me noticing, and so much was moved around that I really can't keep track of it all. I don't see anything wrong, I just had a couple of questions, but please try to keep them smaller in the future.

File: src/components/Gradebook/GradebookFilters/AssignmentGradeFilter/test.jsx [hunk 4]
```diff
@@ -10,77 +11,70 @@ import {
...
 
-describe('AssignmentGradeFilter', () => {
-  let props = {
-    filterValues: {
-      assignmentGradeMin: '1',
-      assignmentGradeMax: '100',
-    },
-    courseId: '12345',
+jest.mock('data/selectors', () => ({
+  __esModule: true,
```
[the remark this answers is not in the document]
Comment (muselesscreator) [reply src/components/Gradebook/GradebookFilters/AssignmentGradeFilter/test.jsx:15]: necessary to mock an es6 imported module

File: src/components/Gradebook/GradebookTable/LabelReplacements.jsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import {
+  Icon,
+  OverlayTrigger,
+  Tooltip,
+} from '@edx/paragon';
+
+import { TOTAL_COURSE_GRADE_HEADING } from 'data/constants/grades';
+
+const TotalGradeLabelReplacement = () => {
+  const totalGradePercentageMessage = 'Total Grade values are always displayed as a percentage.';
```
[the remark this answers is not in the document]
Comment (muselesscreator) [reply src/components/Gradebook/GradebookTable/LabelReplacements.jsx:14]: honestly, the location for this should be in a messages file, which is upcoming.

File: src/components/Gradebook/SpinnerIcon.jsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable react/sort-comp, react/button-has-type, import/no-named-as-default */
```
[the remark this answers is not in the document]
Comment (muselesscreator) [reply src/components/Gradebook/SpinnerIcon.jsx:1]: some, maybe.  But I am of the opinion that we should disable some at the app-level

File: src/components/Gradebook/index.jsx [hunk 7]
```diff
@@ -12,7 +10,6 @@ import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
 import { faFilter } from '@fortawesome/free-solid-svg-icons';
 import PageButtons from '../PageButtons';
 import Drawer from '../Drawer';
-import initialFilters from '../../data/constants/filters';
 import ConnectedFilterBadges from '../FilterBadges';
```
[the remark this answers is not in the document]
Comment (muselesscreator) [reply src/components/Gradebook/index.jsx:13]: probably

File: src/data/reducers/roles.js (hunk 3 above)
Comment (muselesscreator) [reply src/data/reducers/roles.js:4]: because there were complaints that the recent update caused a brief flash of "you are not allowed" until the roles fetch came back.  This way, you default to "fetching" which makes more sense anyway.

File: src/components/Gradebook/GradebookTable/index.jsx [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  formatHeadings = () => (
+    this.props.headings.length
+      ? this.props.headings.map(this.replaceHeader)
+      : this.props.headings
+  )
+
+  roundGrade = percent => parseFloat((percent || 0).toFixed(DECIMAL_PRECISION));
+
+  data = () => this.props.grades.map(entry => ({
+    [USERNAME_HEADING]: (<Fields.Username entry={entry} />),
```
[the remark this answers is not in the document]
Comment (muselesscreator) [reply src/components/Gradebook/GradebookTable/index.jsx:41]: mostly because if I passed username I would also have to pass external_user_key and this felt cleaner

File: src/components/Gradebook/GradebookTable/GradeButton.jsx [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    });
+  }
+
+  render() {
+    return (
+      <Button
+        variant="link"
+        className={classNames(
+          'btn-header',
+          { 'grade-button': this.props.variant === VARIANTS.percent },
```
[the remark this answers is not in the document]
Comment (muselesscreator) [reply src/components/Gradebook/GradebookTable/GradeButton.jsx:36]: no good reason.  This was mostly just trying to compile the original 2 components into 1.  But you are correct that a flag probably makes more sense.

File: src/components/Gradebook/index.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (muselesscreator) [reply src/components/Gradebook/index.jsx:13]: I think I will hold off on that rename until the follow-on PR that is already re-organizing some of that.

File: src/components/Gradebook/GradebookTable/GradeButton.jsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (muselesscreator) [reply src/components/Gradebook/GradebookTable/GradeButton.jsx:36]: uuuuuugh, yes I believe it was.  removing variants
